### ISSUE TITLE
http patch deltas broken by upgrade from 7.5.12 -> 8.04

### ISSUE THREAD

**Issue by dweeden** [OData/AspNetCoreOData#417]
The delta patch parameter to this controller method had changed properties prior to upgrade but their count became zero afterwards

``` csharp
[HttpPatch("odata/[controller]({id:int})")]
public ActionResult Patch(int id, [FromBody] Delta<T> patch)
{
	if (!ModelState.IsValid)
	{
		return BadRequest(ModelState);
	}

	var result = _entities.SingleOrDefault(e => e.Id == id);

	if (result == null)
	{
		return NotFound();
	}

	// update entity and save
	patch.Patch(result);
	_context.SaveChanges();

	return NoContent();
}
````

**Comment by xuzhg** [comment]
@dweeden can you share with me a repro?

**Comment by dweeden** [comment]
sorry this took awhile, i'm fairly useless with git, "initial commit" has this at 7.5.12 and "updated to 8.04" should be self explanatory :)

https://github.com/dweeden/HttpPatchIssue

you should be able to see the before and after behaviour by setting a breakpoint in patch method of the person controller and checking the changed properties after attempting an update in the grid

everything is fine in 7.5.12, not so much in 8.04

**Comment by xuzhg** [comment]
@dweeden This link doesn't work. Would you please check the repository is gone?

**Comment by dweeden** [comment]
sorry, screwed up and deleted it :(  i will try again

**Comment by dweeden** [comment]
Have actually discovered a second issue broken by upgrading to the latest (8.0.5)

In addition to the patch issue, i found that the template routing seems to be broken too :(

I found this out as I have a base class for my odata controllers in my real project so i can have multiple editable grids

I think the commit comments are pretty clear in the following repo but please let me know anything is unclear

both are available in [dweeden/
AspNetCoreODataIssues](https://github.com/dweeden/AspNetCoreODataIssues)

**Comment by dweeden** [comment]
have figured out a workaround but the heart of the problem appears to be [controller] no longer being an adequate differentiator between controller methods as it was in 7.x

please see latest commits in new project where I made code changes as follows to get something working with 8.06

* critical get method not populating id param problem side stepped by commenting out method that takes a param
* http patch issue sidestepped by lifting attribute out to deriving class to avoid [controller] not being handled as expected
* note that [controller] differentiated delete is handled properly despite runtime warning. ie.

**Comment by julealgon** [comment]
> Have figured out a workaround but the heart of the problem appears to be that using [controller] within a route template is no longer an adequate differentiator between actions across controllers as it was in 7.x, causing definite issues for HTTP GET and PATCH (maybe PUT too?)

@dweeden , please refer to this issue for the token replacement problem. If you have any details to add, please add it there.
- https://github.com/OData/AspNetCoreOData/issues/430

**Comment by julealgon** [comment]
@dweeden , some feedback on your repo:
1. You can use `[ApiController]` and have automatic model state validation take place (so no need for manually checking model state and returning bad request)
2. `[ApiController]` also allows you to pass complex types in the body by default (so no need for `[FromBody]` in the `Delta`, for example)
3. I'd not recommend using `NoContent` for a patch. Instead, prefer `Updated(entity)`, which allows you to optionally return the updated entity using the `Prefer: return=representation` header. This is analogous to `Created(entity)` for OData POST, but for use with `PATCH` and `PUT`.

Having said that, patch seems to be working fine in your repo. This is the test I did:
`PATCH http://localhost:5222/odata/animal(1)`
Body:
```json
{
    "Name": "New"
}
```
Result (after replacing `NoContent` with `Updated` plus the prefer header as above):
```json
{
    "@odata.context": "http://localhost:5222/odata/$metadata#Animal/$entity",
    "AnimalId": 1,
    "Name": "New"
}
```

**Comment by dweeden** [comment]
@julealgon  Thanks for the pointer to the other issue and feedback on the repo.   I'll merge those suggestions into my code.

May even try tinkering with Rodrigo's suggestion (referenced in #430) within a fork of the upstream AspNetCore to see if I can correct it there as that seems the proper place to apply a fix.

**Comment by julealgon** [comment]
Glad to help @dweeden . Assuming patch is working fine after the `[controller]` replacement, would you consider this issue a duplicate of #430 , or do you still see a reason to keep it open?

My understanding is that the lack of replacement was what was causing your initial issue with PATCH, correct me if I'm wrong here.

**Comment by dweeden** [comment]
@julealgon Likely closable as a duplicate considering template replacement not occurring is the root cause but... it was odd that DELETE was kind of handled properly whereas PUT and GET definitely were broken.

Not sure if that indicates maybe something else going on in the specifics of how each verb is being handled?    You and/or @xuzhg would be better suited than me to make that call.

**Comment by julealgon** [comment]
Interesting, I completely skipped your comments on DELETE @dweeden , my bad.

After some testing on my side, I concluded that DELETEs are working because they are being recognized as standard MVC routes only, and not as OData routes. Enabling the route debugger, this is what I get on your API:

![image](https://user-images.githubusercontent.com/461579/150203134-e879e28e-0594-4db7-b722-ece952673b97.png)

This shows us that both DELETE actions are only recognized as non-OData.

As you saw, OData can't recognize the `[controller]` part of the route, which causes the runtime warning. Due to that, the OData attribute routing convention is skipped, and the same action is then inspected by MVC's attribute routing. Since that same route is a valid MVC route, the endpoint is considered a pure MVC route. This is a fallback to MVC and is by design.

Now, if for example you change your `Delete` method to take an `int key` instead of an `int id`, your endpoint will start to be detected by OData _convention_ routing, since it matches the "Delete" template specified here:
- [OData Routing - Built-in Routing Conventions](https://docs.microsoft.com/en-us/odata/webapi/built-in-routing-conventions#built-in-routing-conventions-1)
![image](https://user-images.githubusercontent.com/461579/150207880-096614f6-9403-4b00-87cc-2f17aa1d7d6c.png)

Changing the argument name from `id` to `key` causes the route to be matched by _both_ OData (via convention but _not_ via attribute routing) and MVC (via attribute routing):

![image](https://user-images.githubusercontent.com/461579/150208200-cbe153a1-1c36-42f5-98bd-5aa999ee0207.png)

If you then try to call this DELETE endpoint, it will fail with an exception:
> Microsoft.AspNetCore.Routing.Matching.AmbiguousMatchException: The request matched multiple endpoints. Matches: 
> 
> AspNetCoreODataIssues.Controllers.PersonController.Delete (AspNetCoreODataIssues)
AspNetCoreODataIssues.Controllers.PersonController.Delete (AspNetCoreODataIssues)

This particular problem is documented in 2 places (2 distinct issues affect it):
- https://github.com/OData/AspNetCoreOData/issues/431
- https://github.com/OData/AspNetCoreOData/issues/428

Alternatively, if instead of renaming the argument, we mark your base delete method as `virtual`, and override it like this in the Animal controller:
```csharp
    [HttpDelete("odata/Animal({id:int})")]
    public override IActionResult Delete(int id)
    {
        return base.Delete(id);
    }
```

It then shows up as an OData route (via OData attribute routing now):
![image](https://user-images.githubusercontent.com/461579/150203592-c45508fd-ac40-4e2f-8501-76f5970ef08c.png)

Notice how it is _not_ detected as an MVC route now, because OData attribute routing "won": it only appears once, and also works normally as expected.

I initially thought the issue was caused by the introduction of a base class in the mix, but this is not the case. The behavior I described above works both on a single class as well as a class hierarchy in a consistent (wrong) manner.

TL;DR: your DELETE behavior is related to another issue and not caused necessarily by the replacement problem.

@xuzhg , I'd recommend we close this and handle the issues linked separately.

**Comment by dweeden** [comment]
@julealgon  Great analysis.  Sounds closable as duplicate.
